internal/xds/balancer/clusterimpl/tests/balancer_test.go

```diff
@@ -663,3 +664,218 @@ func (s) TestLRSLogicalDNS(t *testing.T) {
+		Listener:   lis,
+		EmptyCallF: func(context.Context, *testpb.Empty) (*testpb.Empty, error) { return &testpb.Empty{}, nil },
+	})
+	defer server.Stop()
+	host, port := hostAndPortFromAddress(t, server.Address)
+
+	const (
+		clusterName       = "test-xds-service"
+		logicalDNSCluster = "logical_dns_cluster"
+		virtualHostName   = "test-route"
+	)
+
+	// Configure xDS resources.
+	ldnsCluster := e2e.ClusterResourceWithOptions(e2e.ClusterOptions{
+		Type:        e2e.ClusterTypeLogicalDNS,
+		ClusterName: logicalDNSCluster,
+		DNSHostName: host,
+		DNSPort:     port,
+	})
+	cluster := e2e.ClusterResourceWithOptions(e2e.ClusterOptions{
+		ClusterName: clusterName,
+		Type:        e2e.ClusterTypeAggregate,
+		ChildNames:  []string{logicalDNSCluster},
+	})
+	route := e2e.DefaultRouteConfig("test-route", virtualHostName, clusterName)
+	listener := e2e.DefaultClientListener(virtualHostName, route.Name)
+	updateOpts := e2e.UpdateOptions{
+		NodeID:         nodeID,
+		Endpoints:      nil,
+		Clusters:       []*v3clusterpb.Cluster{cluster, ldnsCluster},
+		Routes:         []*v3routepb.RouteConfiguration{route},
+		Listeners:      []*v3listenerpb.Listener{listener},
+		SkipValidation: true,
+	}
+
+	// Replace DNS resolver with a wrapped resolver to capture ResolveNow calls.
+	resolveNowCh := make(chan struct{}, 1)
+	dnsR := manual.NewBuilderWithScheme("dns")
+	dnsResolverBuilder := resolver.Get("dns")
+	resolver.Register(dnsR)
+	t.Cleanup(func() { resolver.Register(dnsResolverBuilder) })
+	dnsR.ResolveNowCallback = func(resolver.ResolveNowOptions) {
+		close(resolveNowCh)
+	}
+	dnsR.UpdateState(resolver.State{Addresses: []resolver.Address{{Addr: fmt.Sprintf("%s:%d", host, port)}}})
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	if err := mgmtServer.Update(ctx, updateOpts); err != nil {
+		t.Fatalf("Failed to update xDS resources: %v", err)
+	}
+
+	conn, err := grpc.NewClient("xds:///test-route", grpc.WithResolvers(resolverBuilder), grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		t.Fatalf("Failed to create client: %v", err)
+	}
+	defer conn.Close()
+
+	conn.Connect()
+	client := testgrpc.NewTestServiceClient(conn)
+
+	// Verify initial RPC routes correctly to backend.
+	testutils.AwaitState(ctx, t, conn, connectivity.Ready)
+	if _, err := client.EmptyCall(ctx, &testpb.Empty{}); err != nil {
+		t.Fatalf("client.EmptyCall() failed: %v", err)
+	}
+
+	// Stopping the server listener will close the transport on the client,
+	// which will lead to the channel eventually moving to IDLE.
+	lis.Stop()
+	testutils.AwaitState(ctx, t, conn, connectivity.Idle)
+
+	// Make an RPC to reconnect and thereby move to TRANSIENT_FAILURE upon
+	// connection failure.
+	client.EmptyCall(ctx, &testpb.Empty{})
+	testutils.AwaitState(ctx, t, conn, connectivity.TransientFailure)
+
+	// An RPC at this point is expected to fail.
+	if _, err = client.EmptyCall(ctx, &testpb.Empty{}); err == nil {
```

Comment:
Nit: Instead of checking for `err == nil`, could you please check for status of `Unavailable` which is what you would expect the channel is in TF.